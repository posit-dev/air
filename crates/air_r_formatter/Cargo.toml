
[package]
authors.workspace = true
categories.workspace = true
description = "Air R Formatter"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "air_r_formatter"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
air_r_syntax = { workspace = true }
biome_formatter = { workspace = true }
biome_rowan = { workspace = true }
itertools = { workspace = true }
settings = { workspace = true, features = ["biome", "serde"] }
tracing = { workspace = true }

[dev-dependencies]
air_formatter_test = { workspace = true }
air_r_parser = { workspace = true }
biome_parser = { workspace = true }
line_ending = { workspace = true }
tests_macros = { workspace = true }

[lints]
workspace = true
