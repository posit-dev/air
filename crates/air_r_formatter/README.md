To run tests you need to install [cargo insta](https://crates.io/crates/cargo-insta):

```sh
cargo install cargo-insta
```

Run and review tests with:

```sh
cargo insta test
cargo insta review
```

The `src/generated.rs` file is produced by running this at the workspace root:

```sh
just gen-formatter
```
