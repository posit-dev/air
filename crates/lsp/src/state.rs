use std::collections::HashMap;

use anyhow::anyhow;
use url::Url;

use crate::documents::Document;

#[derive(Clone, Default, Debug)]
/// The world state, i.e. all the inputs necessary for analysing or refactoring
/// code. This is a pure value. There is no interior mutability in this data
/// structure. It can be cloned and safely sent to other threads.
pub(crate) struct WorldState {
    /// Watched documents
    pub(crate) documents: HashMap<Url, Document>,

    /// The scopes for the console. This currently contains a list (outer `Vec`)
    /// of names (inner `Vec`) within the environments on the search path, starting
    /// from the global environment and ending with the base package. Eventually
    /// this might also be populated with the scope for the current environment
    /// in debug sessions (not implemented yet).
    ///
    /// This is currently one of the main sources of known symbols for
    /// diagnostics. In the future we should better delineate interactive
    /// contexts (e.g. the console, but scripts might also be treated as
    /// interactive, which could be a user setting) and non-interactive ones
    /// (e.g. a package). In non-interactive contexts, the lexical scopes
    /// examined for diagnostics should be fully determined by variable bindings
    /// and imports (code-first diagnostics).
    ///
    /// In the future this should probably become more complex with a list of
    /// either symbol names (as is now the case) or named environments, such as
    /// `pkg:ggplot2`. Storing named environments here will allow the LSP to
    /// retrieve the symbols in a pull fashion (the whole console scopes are
    /// currently pushed to the LSP), and cache the symbols with Salsa. The
    /// performance is not currently an issue but this could change once we do
    /// more analysis of symbols in the search path.
    #[allow(dead_code)]
    pub(crate) console_scopes: Vec<Vec<String>>,

    /// Currently installed packages
    #[allow(dead_code)]
    pub(crate) installed_packages: Vec<String>,
}

impl WorldState {
    pub(crate) fn get_document(&self, uri: &Url) -> Option<&Document> {
        self.documents.get(uri)
    }

    pub(crate) fn get_document_mut(&mut self, uri: &Url) -> Option<&mut Document> {
        self.documents.get_mut(uri)
    }

    pub(crate) fn get_document_or_error(&self, uri: &Url) -> anyhow::Result<&Document> {
        match self.get_document(uri) {
            Some(doc) => Ok(doc),
            None => Err(anyhow!("Can't find document for URI {uri}")),
        }
    }

    pub(crate) fn get_document_mut_or_error(&mut self, uri: &Url) -> anyhow::Result<&mut Document> {
        match self.get_document_mut(uri) {
            Some(doc) => Ok(doc),
            None => Err(anyhow!("Can't find document for URI {uri}")),
        }
    }

    pub(crate) fn workspace_uris(&self) -> Vec<Url> {
        self.documents.iter().map(|elt| elt.0.clone()).collect()
    }
}
