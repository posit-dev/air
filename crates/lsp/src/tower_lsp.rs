//
// tower_lsp.rs
//
// Copyright (C) 2022-2024 Posit Software, PBC. All rights reserved.
//
//

#![allow(deprecated)]

use strum::IntoStaticStr;

use tokio::io::{AsyncRead, AsyncWrite};
use tokio::sync::mpsc::unbounded_channel as tokio_unbounded_channel;
use tower_lsp::Client;
use tower_lsp::LanguageServer;
use tower_lsp::LspService;
use tower_lsp::jsonrpc::Result;
use tower_lsp::lsp_types::*;
use tower_lsp::{ClientSocket, jsonrpc};

use crate::handlers_ext::ViewFileParams;
use crate::main_loop::Event;
use crate::main_loop::GlobalState;
use crate::main_loop::TokioUnboundedSender;

// Based on https://stackoverflow.com/a/69324393/1725177
macro_rules! cast_response {
    ($target:expr, $pat:path) => {{
        match $target {
            Ok($pat(resp)) => Ok(resp),
            Err(err) => Err(new_jsonrpc_error(format!("{err:?}"))),
            _ => panic!("Unexpected variant while casting to {}", stringify!($pat)),
        }
    }};
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
pub(crate) enum LspMessage {
    Notification(LspNotification),
    Request(
        LspRequest,
        TokioUnboundedSender<anyhow::Result<LspResponse>>,
    ),
}

#[derive(Debug, IntoStaticStr)]
pub(crate) enum LspNotification {
    Initialized(InitializedParams),
    DidChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams),
    DidChangeConfiguration(DidChangeConfigurationParams),
    DidChangeWatchedFiles(DidChangeWatchedFilesParams),
    DidOpenTextDocument(DidOpenTextDocumentParams),
    DidChangeTextDocument(DidChangeTextDocumentParams),
    DidSaveTextDocument(DidSaveTextDocumentParams),
    DidCloseTextDocument(DidCloseTextDocumentParams),
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, IntoStaticStr)]
pub(crate) enum LspRequest {
    Initialize(InitializeParams),
    DocumentFormatting(DocumentFormattingParams),
    Shutdown,
    DocumentRangeFormatting(DocumentRangeFormattingParams),
    AirViewFile(ViewFileParams),
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, IntoStaticStr)]
pub(crate) enum LspResponse {
    Initialize(InitializeResult),
    DocumentFormatting(Option<Vec<TextEdit>>),
    DocumentRangeFormatting(Option<Vec<TextEdit>>),
    Shutdown(()),
    AirViewFile(String),
}

impl std::fmt::Display for LspNotification {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.into())
    }
}
impl std::fmt::Display for LspRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.into())
    }
}
impl std::fmt::Display for LspResponse {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.into())
    }
}

impl LspNotification {
    fn trace(&self) -> TraceLspNotification<'_> {
        TraceLspNotification { inner: self }
    }
}
impl LspRequest {
    fn trace(&self) -> TraceLspRequest<'_> {
        TraceLspRequest { inner: self }
    }
}
impl LspResponse {
    fn trace(&self) -> TraceLspResponse<'_> {
        TraceLspResponse { inner: self }
    }
}

struct TraceLspNotification<'a> {
    inner: &'a LspNotification,
}
struct TraceLspRequest<'a> {
    inner: &'a LspRequest,
}
struct TraceLspResponse<'a> {
    inner: &'a LspResponse,
}

impl std::fmt::Debug for TraceLspNotification<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.inner {
            LspNotification::DidOpenTextDocument(params) => {
                // Ignore the document itself in trace logs
                f.debug_tuple(self.inner.into())
                    .field(&params.text_document.uri)
                    .field(&params.text_document.version)
                    .field(&params.text_document.language_id)
                    .finish()
            }
            _ => std::fmt::Debug::fmt(self.inner, f),
        }
    }
}

impl std::fmt::Debug for TraceLspRequest<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Debug::fmt(self.inner, f)
    }
}

impl std::fmt::Debug for TraceLspResponse<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Debug::fmt(self.inner, f)
    }
}

#[derive(Debug)]
struct Backend {
    /// Channel for communication with the main loop.
    events_tx: TokioUnboundedSender<Event>,

    /// Handle to main loop. Drop it to cancel the loop, all associated tasks,
    /// and drop all owned state.
    _main_loop: tokio::task::JoinSet<()>,
}

impl Backend {
    async fn request(&self, request: LspRequest) -> anyhow::Result<LspResponse> {
        tracing::info!("Incoming: {request}");
        tracing::trace!("Incoming (debug):\n{request:#?}", request = request.trace());

        let (response_tx, mut response_rx) =
            tokio_unbounded_channel::<anyhow::Result<LspResponse>>();

        // Relay request to main loop
        self.events_tx
            .send(Event::Lsp(LspMessage::Request(request, response_tx)))
            .unwrap();

        // Wait for response from main loop
        let response = response_rx.recv().await.unwrap()?;

        tracing::info!("Outgoing: {response}");
        tracing::trace!(
            "Outgoing (debug):\n{response:#?}",
            response = response.trace()
        );
        Ok(response)
    }

    fn notify(&self, notif: LspNotification) {
        tracing::info!("Incoming: {notif}");
        tracing::trace!("Incoming (debug):\n{notif:#?}", notif = notif.trace());

        // Relay notification to main loop
        self.events_tx
            .send(Event::Lsp(LspMessage::Notification(notif)))
            .unwrap();
    }

    async fn air_view_file(&self, params: ViewFileParams) -> tower_lsp::jsonrpc::Result<String> {
        cast_response!(
            self.request(LspRequest::AirViewFile(params)).await,
            LspResponse::AirViewFile
        )
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for Backend {
    async fn initialize(&self, params: InitializeParams) -> Result<InitializeResult> {
        cast_response!(
            self.request(LspRequest::Initialize(params)).await,
            LspResponse::Initialize
        )
    }

    async fn initialized(&self, params: InitializedParams) {
        self.notify(LspNotification::Initialized(params));
    }

    async fn shutdown(&self) -> Result<()> {
        cast_response!(
            self.request(LspRequest::Shutdown).await,
            LspResponse::Shutdown
        )
    }

    async fn did_change_workspace_folders(&self, params: DidChangeWorkspaceFoldersParams) {
        self.notify(LspNotification::DidChangeWorkspaceFolders(params));
    }

    async fn did_change_configuration(&self, params: DidChangeConfigurationParams) {
        self.notify(LspNotification::DidChangeConfiguration(params));
    }

    async fn did_change_watched_files(&self, params: DidChangeWatchedFilesParams) {
        self.notify(LspNotification::DidChangeWatchedFiles(params));
    }

    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        self.notify(LspNotification::DidOpenTextDocument(params));
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        self.notify(LspNotification::DidChangeTextDocument(params));
    }

    async fn did_save(&self, params: DidSaveTextDocumentParams) {
        self.notify(LspNotification::DidSaveTextDocument(params));
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        self.notify(LspNotification::DidCloseTextDocument(params));
    }

    async fn formatting(&self, params: DocumentFormattingParams) -> Result<Option<Vec<TextEdit>>> {
        cast_response!(
            self.request(LspRequest::DocumentFormatting(params)).await,
            LspResponse::DocumentFormatting
        )
    }

    async fn range_formatting(
        &self,
        params: DocumentRangeFormattingParams,
    ) -> Result<Option<Vec<TextEdit>>> {
        cast_response!(
            self.request(LspRequest::DocumentRangeFormatting(params))
                .await,
            LspResponse::DocumentRangeFormatting
        )
    }
}

pub async fn start_lsp<I, O>(read: I, write: O)
where
    I: AsyncRead + Unpin,
    O: AsyncWrite,
{
    let (service, socket) = new_lsp();
    let server = tower_lsp::Server::new(read, write, socket);
    server.serve(service).await;
}

fn new_lsp() -> (LspService<Backend>, ClientSocket) {
    let init = |client: Client| {
        let (state, events_tx) = GlobalState::new(client);

        // Start main loop and hold onto the handle that keeps it alive
        let main_loop = state.start();

        Backend {
            events_tx,
            _main_loop: main_loop,
        }
    };

    LspService::build(init)
        .custom_method("air/viewFile", Backend::air_view_file)
        .finish()
}

fn new_jsonrpc_error(message: String) -> jsonrpc::Error {
    jsonrpc::Error {
        code: jsonrpc::ErrorCode::ServerError(-1),
        message: message.into(),
        data: None,
    }
}
