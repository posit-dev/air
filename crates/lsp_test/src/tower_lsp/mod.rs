pub(crate) mod codec;

#[allow(dead_code)]
pub(crate) mod request;
