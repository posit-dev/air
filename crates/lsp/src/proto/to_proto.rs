//
// to_proto.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

// Utilites for converting internal types to LSP types

use anyhow::Context;
use biome_line_index::LineIndex;
use biome_text_size::TextRange;
use biome_text_size::TextSize;
use settings::LineEnding;
use tower_lsp::lsp_types;

use crate::proto::PositionEncoding;
use crate::text_edit::Indel;
use crate::text_edit::TextEdit;

/// The function is used to convert TextSize to a LSP position.
/// From `biome_lsp_converters::to_proto::position()`.
pub(crate) fn position(
    offset: TextSize,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
) -> anyhow::Result<lsp_types::Position> {
    let line_col = line_index
        .line_col(offset)
        .with_context(|| format!("Could not convert offset {offset:?} into a line-column index"))?;

    let position = match position_encoding {
        PositionEncoding::Utf8 => lsp_types::Position::new(line_col.line, line_col.col),
        PositionEncoding::Wide(enc) => {
            let line_col = line_index
                .to_wide(enc, line_col)
                .with_context(|| format!("Could not convert {line_col:?} into wide line column"))?;
            lsp_types::Position::new(line_col.line, line_col.col)
        }
    };

    Ok(position)
}

/// The function is used to convert TextRange to a LSP range.
/// From `biome_lsp_converters::to_proto::range()`.
pub(crate) fn range(
    range: TextRange,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
) -> anyhow::Result<lsp_types::Range> {
    let start = position(range.start(), line_index, position_encoding)?;
    let end = position(range.end(), line_index, position_encoding)?;
    Ok(lsp_types::Range::new(start, end))
}

pub(crate) fn text_edit(
    indel: Indel,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
    endings: LineEnding,
) -> anyhow::Result<lsp_types::TextEdit> {
    let range = range(indel.delete, line_index, position_encoding)?;
    let new_text = match endings {
        LineEnding::Lf => indel.insert,
        LineEnding::Crlf => indel.insert.replace('\n', "\r\n"),
    };
    Ok(lsp_types::TextEdit { range, new_text })
}

pub(crate) fn text_edit_vec(
    text_edit: TextEdit,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
    endings: LineEnding,
) -> anyhow::Result<Vec<lsp_types::TextEdit>> {
    text_edit
        .into_iter()
        .map(|indel| self::text_edit(indel, line_index, position_encoding, endings))
        .collect()
}

#[cfg(test)]
pub(crate) fn doc_edit_vec(
    text_edit: TextEdit,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
    endings: LineEnding,
) -> anyhow::Result<Vec<lsp_types::TextDocumentContentChangeEvent>> {
    let edits = text_edit_vec(text_edit, line_index, position_encoding, endings)?;

    Ok(edits
        .into_iter()
        .map(|edit| lsp_types::TextDocumentContentChangeEvent {
            range: Some(edit.range),
            range_length: None,
            text: edit.new_text,
        })
        .collect())
}

pub(crate) fn replace_range_edit(
    range: TextRange,
    replace_with: String,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
    endings: LineEnding,
) -> anyhow::Result<Vec<lsp_types::TextEdit>> {
    let edit = TextEdit::replace(range, replace_with);
    text_edit_vec(edit, line_index, position_encoding, endings)
}

pub(crate) fn replace_all_edit(
    text: &str,
    replace_with: &str,
    line_index: &LineIndex,
    position_encoding: PositionEncoding,
    endings: LineEnding,
) -> anyhow::Result<Vec<lsp_types::TextEdit>> {
    let edit = crate::diff::diff(text, replace_with);
    text_edit_vec(edit, line_index, position_encoding, endings)
}
