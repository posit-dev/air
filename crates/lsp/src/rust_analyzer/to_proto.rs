// --- source
// authors = ["rust-analyzer team"]
// license = "MIT OR Apache-2.0"
// origin = "https://github.com/rust-lang/rust-analyzer/blob/master/crates/rust-analyzer/src/lsp/to_proto.rs"
// ---

//! Conversion of rust-analyzer specific types to lsp_types equivalents.

use super::{
    line_index::LineIndex,
    text_edit::{Indel, TextEdit},
};
use line_ending::LineEnding;
use tower_lsp::lsp_types;

pub(crate) fn text_edit(
    line_index: &LineIndex,
    indel: Indel,
) -> anyhow::Result<lsp_types::TextEdit> {
    let range = biome_lsp_converters::to_proto::range(
        &line_index.index,
        indel.delete,
        line_index.encoding,
    )?;
    let new_text = match line_index.endings {
        LineEnding::Lf => indel.insert,
        LineEnding::Crlf => indel.insert.replace('\n', "\r\n"),
    };
    Ok(lsp_types::TextEdit { range, new_text })
}

pub(crate) fn completion_text_edit(
    line_index: &LineIndex,
    insert_replace_support: Option<lsp_types::Position>,
    indel: Indel,
) -> anyhow::Result<lsp_types::CompletionTextEdit> {
    let text_edit = text_edit(line_index, indel)?;
    Ok(match insert_replace_support {
        Some(cursor_pos) => lsp_types::InsertReplaceEdit {
            new_text: text_edit.new_text,
            insert: lsp_types::Range {
                start: text_edit.range.start,
                end: cursor_pos,
            },
            replace: text_edit.range,
        }
        .into(),
        None => text_edit.into(),
    })
}

pub(crate) fn text_edit_vec(
    line_index: &LineIndex,
    text_edit: TextEdit,
) -> anyhow::Result<Vec<lsp_types::TextEdit>> {
    text_edit
        .into_iter()
        .map(|indel| self::text_edit(line_index, indel))
        .collect()
}
