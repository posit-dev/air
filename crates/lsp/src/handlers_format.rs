//
// handlers_format.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use air_r_formatter::{context::RFormatOptions, format_node};
use biome_formatter::{IndentStyle, LineWidth};
use tower_lsp::lsp_types;

use crate::state::WorldState;
use crate::to_proto;

#[tracing::instrument(level = "info", skip_all)]
pub(crate) fn document_formatting(
    params: lsp_types::DocumentFormattingParams,
    state: &WorldState,
) -> anyhow::Result<Option<Vec<lsp_types::TextEdit>>> {
    let doc = state.get_document(&params.text_document.uri)?;

    let line_width = LineWidth::try_from(80).map_err(|err| anyhow::anyhow!("{err}"))?;

    // TODO: Handle FormattingOptions
    let options = RFormatOptions::default()
        .with_indent_style(IndentStyle::Space)
        .with_line_width(line_width);

    if doc.parse.has_errors() {
        return Err(anyhow::anyhow!("Can't format when there are parse errors."));
    }

    let formatted = format_node(options.clone(), &doc.parse.syntax())?;
    let output = formatted.print()?.into_code();

    // Do we need to check that `doc` is indeed an R file? What about special
    // files that don't have extensions like `NAMESPACE`, do we hard-code a
    // list? What about unnamed temporary files?

    let edits = to_proto::replace_all_edit(&doc.line_index, &doc.contents, &output)?;
    Ok(Some(edits))
}
