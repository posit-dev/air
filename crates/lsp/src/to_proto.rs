//
// to_proto.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

// Utilites for converting internal types to LSP types

pub(crate) use rust_analyzer::to_proto::text_edit_vec;

#[cfg(test)]
pub(crate) use biome_lsp_converters::to_proto::range;

use crate::rust_analyzer::{self, line_index::LineIndex, text_edit::TextEdit};
use biome_text_size::TextRange;
use tower_lsp::lsp_types;

pub(crate) fn doc_edit_vec(
    line_index: &LineIndex,
    text_edit: TextEdit,
) -> anyhow::Result<Vec<lsp_types::TextDocumentContentChangeEvent>> {
    let edits = text_edit_vec(line_index, text_edit)?;

    Ok(edits
        .into_iter()
        .map(|edit| lsp_types::TextDocumentContentChangeEvent {
            range: Some(edit.range),
            range_length: None,
            text: edit.new_text,
        })
        .collect())
}

pub(crate) fn replace_range_edit(
    line_index: &LineIndex,
    range: TextRange,
    replace_with: String,
) -> anyhow::Result<Vec<lsp_types::TextEdit>> {
    let edit = TextEdit::replace(range, replace_with);
    text_edit_vec(line_index, edit)
}

pub(crate) fn replace_all_edit(
    line_index: &LineIndex,
    text: &str,
    replace_with: &str,
) -> anyhow::Result<Vec<lsp_types::TextEdit>> {
    let edit = TextEdit::diff(text, replace_with);
    text_edit_vec(line_index, edit)
}
