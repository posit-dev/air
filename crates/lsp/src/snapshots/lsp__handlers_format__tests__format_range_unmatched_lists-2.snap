---
source: crates/lsp/src/handlers_format.rs
expression: output2
---
0+0
1 + 1
{
	2 + 2
}
3+3
